sys/kern/ringbuf.c

```diff
@@ -28,6 +29,23 @@ static void consume(ringbuf_t *buf, unsigned bytes) {
     buf->tail = 0;
 }
 
+/* below two function are to help to track ringbuf pointers in case when it is
+ * changed by something else. e.q. DMA */
+void ringbuf_produce(ringbuf_t *buf, unsigned bytes) {
+  assert(buf->count <= buf->size && bytes <= buf->size &&
+         buf->count <= buf->size - bytes);
+  buf->count += bytes;
+  buf->head += bytes;
+  buf->head %= buf->size;
```

Comment:
Modulo operation can be slow here. Since `buf->head` won't be bigger than `2 * buf->size` it can be done with conditional subtraction:
```
if (buf->head >= buf->size)
    buf->head -= buf->size;
```

Reply:
Thank you - Done